conversion/resource_generation.go

```diff
@@ -0,0 +1,413 @@
+	if err != nil {
+		return region, err
+	}
+	for _, replica := range instanceConfig.Replicas {
+		if replica.DefaultLeaderLocation {
+			region = replica.Location
+		}
+	}
+	return region, nil
+}
+
+// Returns connection profiles for a given region
+func (r *ResourceGenerationStruct) getConnProfilesRegion(ctx context.Context, projectId string, region string, dsClient *datastream.Client) ([]string, error) {
+	profilesIt := dsClient.ListConnectionProfiles(ctx, &datastreampb.ListConnectionProfilesRequest{Parent: "projects/" + projectId + "/locations/" + region})
+	var profiles []string = []string{}
+	for {
+		resp, err := profilesIt.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return profiles, err
+		} else {
+			profiles = append(profiles, strings.Split(resp.Name, "/")[5])
+		}
+	}
+	return profiles, nil
+}
+
+// returns true if connection profile exists else false
+func (r *ResourceGenerationStruct) connectionProfileExists(ctx context.Context, projectId string, profileName string, profileLocation string, connectionProfiles map[string][]string, dsClient *datastream.Client, s resourceGenerationInterface) (bool, error) {
+	// Check if connection profiles for the given region are fetched. if not, fetch them
+	profiles, ok := connectionProfiles[profileLocation]
+	var err error = nil
+	if !ok {
+		profiles, err = s.getConnProfilesRegion(ctx, projectId, profileLocation, dsClient)
+		if err != nil {
+			return false, err
+		}
+		connectionProfiles[profileLocation] = profiles
+	}
+
+	// Check if connection profile exists in the provided region
+	for _, element := range profiles {
+		if element == profileName {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// If any of the resource creation fails, deletes all resources that were created
+func (r ResourceGenerationStruct) connectionProfileCleanUp(ctx context.Context, profiles []*ConnectionProfileReq) error {
+	dsClient := GetDatastreamClient(ctx)
+	for _, profile := range profiles {
+		op, err := dsClient.DeleteConnectionProfile(ctx, &datastreampb.DeleteConnectionProfileRequest{
+			Name: fmt.Sprintf("projects/%s/locations/%s/connectionProfiles/%s", profile.ConnectionProfile.ProjectId, profile.ConnectionProfile.Region, profile.ConnectionProfile.Id),
+		})
+
+		if err != nil {
+			return err
+		}
+
+		err = op.Wait(ctx)
+		if err != nil {
+			return err
+		}
+
+		if profile.ConnectionProfile.BucketName != "" {
+			gcsClient, err := storage.NewClient(ctx)
+
+			if err != nil {
+				return err
+			}
+
+			if err := gcsClient.Bucket(profile.ConnectionProfile.BucketName).Delete(ctx); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Returns source and destination connection profiles to be created
+func (r *ResourceGenerationStruct) getResourcesForCreation(ctx context.Context, projectId string, sourceProfile profiles.SourceProfile, region string, validateOnly bool, dsClient *datastream.Client) ([]*ConnectionProfileReq, []*ConnectionProfileReq, error) {
```

Comment:
Very long method - please refactor.

Suggested refactoring - source and destination connection profiles in separate methods.